encoding/encoding.mbt

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///|
+/// Encode a given string to the specified character encoding and returns the resulting bytes.
+///
+/// # Parameters
+///
+/// - `encoding` : The target encoding format.
+/// - `src`: The input string to be encoded.
+///
+/// # Returns
+///
+/// A `bytes` representing the encoded string in the selected format.
+///
+/// # Examples
+///
+/// ```moonbit
+/// let src = "Hello, World!"
+/// let encoded_bytes = encode(UTF8, src)
+/// ```
+pub fn encode(encoding : Encoding, src : String) -> Bytes {
+  // NOTE: special case: MoonBit String are already valid UTF16(LE) bytes
+  match encoding {
+    UTF16 | UTF16LE => return src.to_bytes()
+    _ => ()
+  }
+  let bytes = src.to_bytes()
+  let chars = decode_strict(UTF16LE, bytes)
+  let new_buf = @buffer.T::new(size_hint=bytes.length())
+  let write = match encoding {
+    UTF8 => write_utf8_char
+    UTF16BE => write_utf16be_char
+    _ => abort("unreachable")
+  }
+  for char in chars {
+    // SAFETY: Assume String are always valid UTF16LE
+    write(new_buf, char.unwrap())
+  }
+  new_buf.to_bytes()
+}
+
+///|
+fn write_char(
+  write : (FixedArray[Byte], Int, Char) -> Int
+) -> (@buffer.T, Char) -> Unit {
+  let fixedArr = FixedArray::makei(4, fn { _ => b'\x00' })
+  fn {
+    buf, value => {
+      let len = write(fixedArr, 0, value)
+      let arr = fixedArr.iter().take(len).collect()
+      buf.write_bytes(@bytes.from_array(arr))
+    }
+  }
+}
```

Comment:
This is not efficient for such a hot function. Every character needs to write data to memory at least 4 times:

1. initialize the fixedarray
2. copy some bytes from fixedarray into array
3. copy bytes from array to `Bytes`
4. copy `Bytes` into buffer

This also causes too much memory allocation.

Reply:
Since the buffer is kept as opaque as possible, the currently useful API is only `write_bytes`. I think my updated code can avoid overhead points 1, 2, and 3, but the last point should not be avoided.

> placeholder for snippet

Reply:
> Since the buffer is kept as opaque as possible, the currently useful API is only `write_bytes`. 

There is a API `@buffer.write_byte` so we can also eliminate the copying of bytes:

```
pub fn write_utf8_char(buf : @buffer.T, value : Char) -> Unit {
  let code = value.to_uint()
  match code {
    _..<0x80 => {
      let b0 = ((code & 0x7F) | 0x00).to_byte()
      buf.write_byte(b0)
    }
    _..<0x0800 => {
      let b0 = (((code >> 6) & 0x1F) | 0xC0).to_byte()
      let b1 = ((code & 0x3F) | 0x80).to_byte()
      buf..write_byte(b0)
            .write_byte(b1)
    }
    ..
  }
}
```

It's kind of like the `Show` trait. The `Show::output(Self, @buffer.T)` writes self into the buffer and always has better performance. If the user doesn't care about the copy overhead, they can use `Show::to_string(Self)` -> String. Most of the implementations of `to_string` in core are:

```
impl Show for SomeType with to_string(self) {
  let buf = @buffer.new()
  self.output(buf)
  buf.to_unchecked_string()
}
```